components/ILIAS/Badge/classes/class.ilBadgePersonalTableGUI.php

```diff
@@ -151,8 +157,31 @@ private function getRecords(): array
         $rows = [];
         $a_user_id = $this->user->getId();
 
-        foreach (ilBadgeAssignment::getInstancesByUserId($a_user_id) as $ass) {
-            $badge = new ilBadge($ass->getBadgeId());
+        $parent_obj_ids = [];
+        $identifications = [];
+        $access_cache = [];
+        $ref_id_cache = [];
+
+        /** @var array<int, array{badge: ilBadge, badgeAssignment: ilBadgeAssignment}> $badge_assignments_badge_map */
+        $badge_assignments_badge_map = [];
+        foreach (ilBadgeAssignment::getInstancesByUserId($a_user_id) as $badge_assignment) {
+            $badge = new ilBadge($badge_assignment->getBadgeId());
+            $parent_obj_ids[] = $badge->getParentId();
+            $badge_assignments_badge_map[] = [
+                'badge_assignment' => $badge_assignment,
+                'badge' => $badge
+            ];
+            if ($badge->getImageRid()) {
+                $identifications[] = $badge->getImageRid();
+            }
+        }
+
+        $this->irss->preload($identifications);
+        $this->object_data_cache->preloadObjectCache(array_unique($parent_obj_ids));
+
+        foreach ($badge_assignments_badge_map as $map) {
+            $badge = $map['badge'];
+            $ass = $map['badge_assignment'];
```

Comment:
We should seperate the data retrieval here:

1. In `\ilBadgePersonalTableGUI::getRecords` we should only fetch/determine the data we actual need for identifcation and sorting, which is (IMO) `id`, `title_sortable`, `wared_by_sortable`, `badge_issued_on`, 'active`. Furthermore, you can store the raw assignment and badge.
2. We then need a change in `\ilBadgePersonalTableGUI::getRows` which a) Determines the `rid` strings from the badges of the actual "page" being presented, and b) enrichtes the `$record` with the UI elements required for the actual presentation:

```php
...
        foreach ($records as $record) {
            yield $row_builder->buildDataRow((string) $record['id'], $this->enrichRecord($record));
        }
...
```